flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java

```diff
@@ -0,0 +1,481 @@
+	// ------------------------------------------------------------------------
+
+	@Override
+	public HashMap<KinesisStreamShard, String> snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
+		if (lastSequenceNums == null) {
+			LOG.debug("snapshotState() requested on not yet opened source; returning null.");
+			return null;
+		}
+
+		if (!running) {
+			LOG.debug("snapshotState() called on closed source; returning null.");
+			return null;
+		}
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Snapshotting state. ...");
+		}
+
+		@SuppressWarnings("unchecked")
+		HashMap<KinesisStreamShard, String> currentSequenceNums = (HashMap<KinesisStreamShard, String>) lastSequenceNums.clone();
+		return currentSequenceNums;
+	}
+
+	@Override
+	public void restoreState(HashMap<KinesisStreamShard, String> restoredState) throws Exception {
+		sequenceNumsToRestore = restoredState;
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
+		if (fetcher == null) {
+			LOG.debug("notifyCheckpointComplete() called on uninitialized source");
+			return;
+		}
+
+		if (!running) {
+			LOG.debug("notifyCheckpointComplete() called on closed source");
+			return;
+		}
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Checkpointing completed for checkpoint id {}", checkpointId);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Miscellaneous utilities
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Utility function to assign shards to a specific consumer task in a round-robin fashion.
+	 */
+	protected static List<KinesisStreamShard> assignShards(List<KinesisStreamShard> shards, int numFlinkConsumerTasks, int thisConsumerTaskIndex) {
+		checkArgument(numFlinkConsumerTasks > 0);
+		checkArgument(thisConsumerTaskIndex < numFlinkConsumerTasks);
+
+		List<KinesisStreamShard> closedShards = new ArrayList<>();
+		List<KinesisStreamShard> openShards = new ArrayList<>();
+
+		for (KinesisStreamShard shard : shards) {
+			if (shard.isClosed()) {
+				closedShards.add(shard);
+			} else {
+				openShards.add(shard);
+			}
+		}
+
+		List<KinesisStreamShard> subscribedShards = new ArrayList<>();
+
+		// separately round-robin assign open and closed shards so that all tasks have a fair chance of being
+		// assigned open shards (set of data records in closed shards are bounded)
+
+		for (int i = 0; i < closedShards.size(); i++) {
+			if (i % numFlinkConsumerTasks == thisConsumerTaskIndex) {
+				subscribedShards.add(closedShards.get(i));
+			}
+		}
+
+		for (int i = 0; i < openShards.size(); i++) {
+			if (i % numFlinkConsumerTasks == thisConsumerTaskIndex) {
+				subscribedShards.add(openShards.get(i));
+			}
+		}
+		return subscribedShards;
+	}
+
+	/**
+	 * Checks that the values specified for config keys in the properties config is recognizable.
+	 */
+	protected static void validatePropertiesConfig(Properties config) {
```

Comment:
How about we check it if the provider type is not set or set to BASIC.
If its set to something else, we just have to trust the user that he knows what he's doing